feat(dhall-docs): add type of expression from source code

german1608 opened the pull request:
As discussed in our call, it could be nice for users to see the type of the file from the source code on the rendered documentation, so this PR does so.

What I'm unsure is _where_ to put the documentation. Currently its on the right side of the breadcrumbs, so its something like:
```
breadcrumbs [ : type ]
```

but that might look bad when the title is really large, for example the `dhall-kubernetes`.

I tried other options as well:

* Another section, with a `<h2>Type</h2>`, but it overloads a little the page
* At right side of "Copy path to clipboard", but it doesn't feel right since the type is for the file, not for "Copy path to clipboard"
* Above the breadcrumbs, but TBH I don't feel it looks great there.

That's all I could think of right now, it's late here so probably I've missed something. If you have another idea, please let me know on the comments

- [x] type on index pages
- [x] handle name shadowing correctly

german1608 commented:
(funny how our best example of something hard to test against is always `dhall-kubernetes`)

german1608 commented:
(as usual, https://hydra.dhall-lang.org/job/dhall-haskell/1898/generate-dhall-docs/latest/download/1/docs)

sjakobi commented on dhall-docs/src/Dhall/Docs/Core.hs:
```diff
@@ -85,6 +88,31 @@ getAllDhallFiles baseDir = do
+        name <- maybeNameInLet expr
+        typeOfName name
+      where
+        -- | Returns the last @e@ in `let x = r in e` let binding returning
+        --   `Nothing` if it isn't a `Var`
```
Is something like this clearer?

> For an expression of the form `let x0 = y0 x1 = y1 ... in e` where `e` is a variable, `maybeNameInLet` returns the variable name.

sjakobi commented on dhall-docs/src/Dhall/Docs/Core.hs:
```diff
@@ -85,6 +88,31 @@ getAllDhallFiles baseDir = do
 
+    extractTypeIfInSource :: Expr Void Import -> Maybe (Expr Void Import)
+    extractTypeIfInSource expr = do
+        name <- maybeNameInLet expr
+        typeOfName name
```
One potential problem with this approach is that we might report the wrong type when we have variable shadowing. For example in

```
let x : T0 = y0
let x : T1 = y1
in x
```

we'd return `T0` when the actual type is `T1`.

I'm actually not sure how to solve this best. We probably have some code in `dhall-lsp-server` that already gets this right. I also think that this might be a good opportunity to try using the `subExpressions` traversal, since it's bottom-up.

german1608 replied:
@sjakobi I haven't checkout `dhall-lsp-server` yet, but wouldn't work if we return the _last_ binding type that matches the name?

sjakobi replied:
> wouldn't work if we return the _last_ binding type that matches the name?

That works when the variable index is `0`. In Dhall, a variable like `x` is really a shorthand for `x@0`. In general, we could also encounter an expression like

```
let x : Bool = True
let x : Natural = 0
in x@1
```

`x@1` refers to he `Bool` `x` here.

I don't really expect that we'll encounter cases like this a lot, but if we do this right already, we might avoid some headaches later on.

german1608 replied:
Alright, and what about searching from the last binding name while the variable index != `0`? and keep decrementing?

I checked out `dhall-lsp-server` but it uses type-checking, and we are here just traversing the original AST from the source code.

german1608 replied:
Alright, and what about searching from the last binding name while the variable index != `0`? and keep decrementing?

I checked out `dhall-lsp-server` but it uses type-checking, and we are here just traversing the original AST from the source code.

sjakobi replied:
> Alright, and what about searching from the last binding name while the variable index != `0`? and keep decrementing?

Sounds good to me!

> I checked out `dhall-lsp-server` but it uses type-checking, and we are here just traversing the original AST from the source code.

Oh, I forgot. Sorry!

german1608 replied:
Didn't noticed I sent the same message twice :laughing:

I pushed the change, please let me know what you think :)

german1608 commented:
> I think the type is especially valuable for navigating a package. Names are vague but types can be can quite expressive! So how about including the type in the index pages, similarly to the Haddock synopsis feature?

I like that idea! That way we can include more info on the index (currently it looks a little bit empty).

german1608 commented:
BTW! I removed the .ttf files since turns out they were unnecessary - the css @font-face stuff already fetches from fonts.google.com database.

More over: they have a cool api that generates the CSS for the fonts so we don't have to worry about maintaining the @font-face stuff:

https://github.com/dhall-lang/dhall-haskell/blob/6e4c52e0deb3e25aee76610bb614b6e5247bd5fe/dhall-docs/src/Dhall/Docs/Html.hs#L251

Last thing: I changed 'Roboto Mono' to 'Fira Code' because the later looks better, its less spacy.

german1608 commented:
The only thing that happens right now is that the type splits over lines if it is large enough:

https://hydra.dhall-lang.org/build/63904/download/1/docs/Natural/index.html

I think that it can be configured by the formatter, I'll take a look at that

sjakobi commented:
I noticed that for quite a few files, the type seems to be missing. For example `take.dhall` in https://hydra.dhall-lang.org/build/63908/download/1/docs/List/index.html.

Any idea why?

german1608 commented:
> I noticed that for quite a few files, the type seems to be missing. For example `take.dhall` in https://hydra.dhall-lang.org/build/63908/download/1/docs/List/index.html.
> 
> Any idea why?

It was a bug in the `getLetBindingsWithName`. It didn't traverse the whole list, but just while the name matched. I'm fixing it.

sjakobi commented on dhall-docs/src/Dhall/Docs/Core.hs:
```diff
@@ -85,6 +88,52 @@ getAllDhallFiles baseDir = do
+            getLetBindingWithIndex index $ getLetBindingsWithName name
+        return exprType
+      where
+        -- | For an expression of the form @let x0 = y0 x1 = y1 ... in e@
```
(My mistake)

```suggestion
        -- | For an expression of the form @let x0 = y0 let x1 = y1 ... in e@
```

sjakobi commented on dhall-docs/src/Dhall/Docs/Core.hs:
```diff
@@ -85,6 +88,52 @@ getAllDhallFiles baseDir = do
 
+    extractTypeIfInSource :: Expr Void Import -> Maybe (Expr Void Import)
+    extractTypeIfInSource expr = do
+        V name index <- maybeNameInLet expr
+        (Binding _ _ _ (Just (_, exprType)) _ _) <-
+            getLetBindingWithIndex index $ getLetBindingsWithName name
+        return exprType
```
I'm somewhat suspect that this could be simplified quite a bit now by using `Dhall.Syntax.multiLet`.

german1608 replied:
Indeed

sjakobi commented on dhall-docs/src/Dhall/Docs/Html.hs:
```diff
@@ -66,6 +78,19 @@ exprToHtml expr = renderTree prettyTree
+        opts :: Pretty.LayoutOptions
+        opts = Pretty.defaultLayoutOptions
+                { Pretty.layoutPageWidth =
+                    Pretty.AvailablePerLine maxLinesForType 1.0
```
This should be simpler for the current usecase:

```suggestion
                    Pretty.Unbounded
```

sjakobi commented on dhall-docs/src/Dhall/Docs/Html.hs:
```diff
@@ -66,6 +78,19 @@ exprToHtml expr = renderTree prettyTree
+    typeLayout :: Pretty.Doc ann -> Pretty.SimpleDocStream ann
+    typeLayout = Pretty.removeTrailingWhitespace . Pretty.layoutSmart opts
+      where
+        -- this is done so the type of a dhall file fits in a single line
```
Laying out all types on a single line is not really ideal. These lines could get _very_ long. I'm fine with improving the layout at a later stage though.

german1608 replied:
I'll add it to my task list to not forget this.

Gabriel439 commented:
Just a quick drive-by comment on the issue of tests since that came up.  An easy way to test this is for the program to generate a pure data structure representing the desired directory tree and then create a separate data structure for writing out the desired directory tree to the filesystem.

german1608 commented:
> Just a few more comments. Nearly done I think.
> 
> I also feel like we're by and by moving into tricky territory where it would be nice to have some tests so we don't easily introduce bugs later on.
> 
> For example finding the right type annotation for some variable…
> 
> I kind of don't want to wait with tests for too long, because at some later stage we might have forgotten what the tricky bits are and will have more difficulty creating good test cases.
> 
> With golden tests like those that you've already set up, updating the expected test output as you add more features should be very simple.
> 
> How do you feel about this, @german1608?

Honestly yes, I'm pretty sure of the code that I'm writing because _I'm_ the one doing it _every_ day, but I know that after GSoC project finish I won't be all-day contributing to the project. I've been thinking a lot of this past weeks, but didn't did anything because I'm a little bit unsure about it. Here are my thoughs about it:

* IMO having a golden test to check how the HTML renders can cause some friction right now since the HTML changes really frequently.
* What we could do is to test the tricky bits that are not related to the final HTML, as you suggested right there. We could have test for things like:
  * Get the type of a "dhall file"
  * Get the assertions in dhall file
  ... and so on
* Following up with @Gabriel439 [comment](https://github.com/dhall-lang/dhall-haskell/pull/1898#issuecomment-652617249): i like the idea, I thought of that earlier but having some trouble implementing it.

I definetly want to have tests for the code soon, so let's do a thing: I can fix your last comments @sjakobi (and any other suggestions you might have @Gabriel439) and my next PR will include:

* Test setup for the whole `dhall-docs` utility, using the idea of @Gabriel439 and adding some other isolated tests for the other tricky bits (e.g. type of a file, assertions of a file).
* Code examples, extracted from the `assert`ions.

... _and_ after that PR, we will have a solid MVP that we can release and receive feedback from the community.

What do you think?

sjakobi commented on dhall-docs/src/Dhall/Docs/Html.hs:
```diff
@@ -66,6 +78,18 @@ exprToHtml expr = renderTree prettyTree
+    typeLayout :: Pretty.Doc ann -> Pretty.SimpleDocStream ann
+    typeLayout = Pretty.removeTrailingWhitespace . Pretty.layoutSmart opts
+      where
+        -- this is done so the type of a dhall file fits in a single line
+        -- its a safe value, since types in source codes are not that large
```
I think this comment is a bit outdated now. At the least the bit regarding the safe vaue.

german1608 commented:
> What kind of friction do you have in mind? Updating the expected test output really should be as simple as running `stack test dhall-docs --test-arguments --accept`.
>
Yes, I know we can do so but then we need to update them really frequently since HTML ui/ux currently changes kind of often, that is the friction I'm thinking about. Maybe on my next PR I'll add them, I'll think about it, UI is a little bit consolidated.
 
> I've got to admit that I don't really understand what Gabriel was suggesting there. As long as you got it, @german1608, I don't mind though!

If I understand correctly, Gabriel suggested that we can divide the code in "pure" and "impure", with the "pure" part only generating the _desired_ docs structure, and the "impure" just doing IO to write files.


> Ultimately, I don't care too much about the format of the tests though. You've been doing great work, @german1608, and I'm sure you'll figure out a good test setup too.

Thanks :smile:!

Gabriel439 commented:
@german1608 @sjakobi: The basic idea I had in mind was to define a type like this:

```
data Tree = Directory (Map Text Tree) | File Text
```

... and then you can split your code into three phases:

```haskell
-- Read in the Dhall package as a `Tree`
input :: FilePath -> IO Tree

-- Transform the Dhall code into HTML.  This is a pure function that you can easily test
document :: Tree -> Either SomeError Tree

-- Write out the HTML to files
output :: FilePath -> Tree -> IO ()
```
